File: generated_codes/chat_agents/chat-agents-cpp/linked_list/claude_sonnet_4.6/output10_linked_list.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>

class ConcurrentDataStructure : public SetADT {
    struct Node {
        int val;
        std::atomic<Node*> next;
        Node(int v) : val(v), next(nullptr) {}
    };

    static bool is_marked_ref(Node* p) {
        return (reinterpret_cast<uintptr_t>(p) & 1UL) != 0;
    }
    static Node* get_unmarked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1UL);
    }
    static Node* get_marked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1UL);
    }

    Node* head;
    Node* tail;

    // Simple pause for contention backoff (no OS sleep — stays lock-free)
    static void backoff(int& delay) {
        for (int i = 0; i < delay; ++i)
            std::atomic_signal_fence(std::memory_order_seq_cst);
        if (delay < 1024) delay *= 2;
    }

    std::pair<Node*, Node*> find(int key) {
        int delay = 1;
    retry:
        Node* pred = head;
        Node* curr = get_unmarked_ref(pred->next.load(std::memory_order_acquire));
        while (true) {
            Node* raw  = curr->next.load(std::memory_order_acquire);
            Node* succ = get_unmarked_ref(raw);
            while (is_marked_ref(raw)) {
                if (!pred->next.compare_exchange_strong(curr, succ,
                        std::memory_order_acq_rel, std::memory_order_relaxed)) {
                    backoff(delay);
                    goto retry;
                }
                delete curr;
                curr = succ;
                raw  = curr->next.load(std::memory_order_acquire);
                succ = get_unmarked_ref(raw);
            }
            if (curr->val >= key) return {pred, curr};
            pred = curr;
            curr = succ;
        }
    }

public:
    ConcurrentDataStructure() {
        head = new Node(INT_MIN);
        tail = new Node(INT_MAX);
        head->next.store(tail, std::memory_order_relaxed);
    }

    ~ConcurrentDataStructure() {
        Node* n = get_unmarked_ref(head->next.load(std::memory_order_relaxed));
        delete head;
        while (n) {
            Node* nx = get_unmarked_ref(n->next.load(std::memory_order_relaxed));
            delete n;
            n = nx;
        }
    }

    bool contains(int key) override {
        Node* curr = get_unmarked_ref(head->next.load(std::memory_order_acquire));
        while (curr->val < key)
            curr = get_unmarked_ref(curr->next.load(std::memory_order_acquire));
        return curr->val == key && !is_marked_ref(curr->next.load(std::memory_order_acquire));
    }

    bool add(int key) override {
        int delay = 1;
        while (true) {
            auto [pred, curr] = find(key);
            if (curr->val == key) return false;
            Node* nd = new Node(key);
            nd->next.store(curr, std::memory_order_relaxed);
            if (pred->next.compare_exchange_strong(curr, nd,
                    std::memory_order_acq_rel, std::memory_order_relaxed))
                return true;
            delete nd;
            backoff(delay);
        }
    }

    bool remove(int key) override {
        int delay = 1;
        while (true) {
            auto [pred, curr] = find(key);
            if (curr->val != key) return false;
            Node* succ = get_unmarked_ref(curr->next.load(std::memory_order_acquire));
            Node* marked = get_marked_ref(succ);
            if (!curr->next.compare_exchange_strong(succ, marked,
                    std::memory_order_acq_rel, std::memory_order_relaxed)) {
                backoff(delay);
                continue;
            }
            pred->next.compare_exchange_strong(curr, succ,
                std::memory_order_acq_rel, std::memory_order_relaxed);
            return true;
        }
    }
};

File: generated_codes/chat_agents/chat-agents-java/linked_list/gemini_java/output10_linked_list.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {

    private static class Node {
        final int key;
        final AtomicMarkableReference<Node> next;

        Node(int key, Node nextNode) {
            this.key = key;
            this.next = new AtomicMarkableReference<>(nextNode, false);
        }
    }

    private static class Window {
        final Node pred;
        final Node curr;

        Window(Node pred, Node curr) {
            this.pred = pred;
            this.curr = curr;
        }
    }

    private final Node head;
    private final Node tail;

    public ConcurrentDataStructure() {
        this.tail = new Node(Integer.MAX_VALUE, null);
        this.head = new Node(Integer.MIN_VALUE, this.tail);
    }

    private Window find(int key) {
        Node pred = null;
        Node curr = null;
        Node succ = null;
        boolean[] marked = {false};
        boolean retry;

        retry_loop:
        while (true) {
            pred = head;
            curr = pred.next.getReference();
            while (true) {
                succ = curr.next.get(marked);
                while (marked[0]) {
                    if (!pred.next.compareAndSet(curr, succ, false, false)) {
                        continue retry_loop;
                    }
                    curr = succ;
                    succ = curr.next.get(marked);
                }
                if (curr.key >= key) {
                    return new Window(pred, curr);
                }
                pred = curr;
                curr = succ;
            }
        }
    }

    @Override
    public boolean add(int key) {
        while (true) {
            Window window = find(key);
            Node pred = window.pred;
            Node curr = window.curr;

            if (curr.key == key) {
                return false;
            } else {
                Node node = new Node(key, curr);
                if (pred.next.compareAndSet(curr, node, false, false)) {
                    return true;
                }
            }
        }
    }

    @Override
    public boolean remove(int key) {
        while (true) {
            Window window = find(key);
            Node pred = window.pred;
            Node curr = window.curr;

            if (curr.key != key) {
                return false;
            } else {
                Node succ = curr.next.getReference();
                if (curr.next.compareAndSet(succ, succ, false, true)) {
                    // Node has been marked
                    pred.next.compareAndSet(curr, succ, false, false);
                    return true;
                }
            }
        }
    }

    @Override
    public boolean contains(int key) {
        boolean[] marked = {false};
        Node curr = head;
        while (curr.key < key) {
            curr = curr.next.getReference();
            curr.next.get(marked);
        }
        return curr.key == key && !marked[0];
    }
}